Avoid re-jitting distributed tasks in the same process

vepadulano opened the pull request.

vepadulano commented:
## Example benchmark of the effects of this PR

### Setup
Hardware

* AMD 5950x (16 physical cores)
* 64 GB RAM
* 1 TB NVMe SSD

Software

* Python 3.11, GCC 12
* dask 2023.2.0

Benchmark

* dimuon tutorial, using distrdf+dask
* 1 file, stored on local SSD. The file is preloaded in filesystem cache before the tests run
* ~2GB of data, 75 TTree clusters
* 2 physical cores. One Dask worker is spawned on each core

The benchmark is run keeping the number of cores fixed, increasing the number of partitions per core. Each test is run 10 times, plots show mean value at each number of partitions and the error is the standard error of the mean.

## Results
### Memory usage

Memory is taken from `psutil.virtual_memory().available`, i.e. the total available memory on the system. The benchmark is the "only" process running on the system (i.e. the machine is disconnected from the internet and no other user applications are running at the same time).

Memory used by the test run is computed as the difference between the memory available before running the analysis and the memory available at the end of the analysis, before exiting the Python script.

![avoid_rejitting_memory](https://user-images.githubusercontent.com/15638895/220674757-66fda652-9b61-4c11-8877-0fd3962833b4.png)

### Runtime

Runtime is computed as the runtime of the analysis, from right before creating the RDataFrame to the moment when the results of the computations are obtained.

![avoid_rejitting_runtime](https://user-images.githubusercontent.com/15638895/220674777-f6756062-73e6-447d-9f91-9b46096180f4.png)

eguiraud commented on bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py:
```diff
@@ -241,3 +241,34 @@ def _(mergeable: ROOT.Detail.RDF.RMergeableVariationsBase, node, backend):
+    """
+    Clone the action held by an RResultPtr, registering it with its RLoopManager
+    """
+    name_with_old_id = snap.filenames[0].removesuffix(".root")
```
what happens if the output file name does not end in `.root`? If it's required that it ends in `.root`, are we checking for that anywhere?

vepadulano replied:
Indeed it should not be an issue, I extended the logic to also accomodate files that do not end in `.root`

eguiraud commented on bindings/experimental/distrdf/python/DistRDF/Ranges.py:
```diff
@@ -14,14 +15,17 @@
     input data source, this can have different attributes.
     """
-    pass
+    rdf_uuid: uuid.UUID
+    id: int
```
need some docs here about what these are and what purpose they serve

eguiraud commented on bindings/experimental/distrdf/python/DistRDF/Ranges.py:
```diff
@@ -39,14 +43,11 @@ class EmptySourceRange(DataRange):
-        self.start = start
-        self.end = end
+    start: int
+    end: int
```
there is still docs about `id` in this class but the data member was moved to `DataRange`

eguiraud commented on tree/dataframe/src/RResultPtr.cxx:
```diff
@@ -14,15 +14,15 @@
 #include <ROOT/RDF/RInterface.hxx>
 #include <ROOT/RDF/RLoopManager.hxx>
 
-using SnapshotPtr_t = ROOT::RDF::RResultPtr<ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager, void>>;
```
Commit messages are to explain the why (and for complex changes the intricacies of the how) that might be lost in the code. It's not clear to me why this change was needed, e.g. if it's just to avoid duplication:

```
[DF] Clean up: remove duplicate type alias declaration
```

vepadulano commented on tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:
```diff
@@ -1647,6 +1647,24 @@ public:
+         for (std::size_t i = 0u; i < sizeof...(ColTypes); ++i)
+            isDef[i] = fIsDefine[i];
+         return isDef;
+      }();
```
I agree that it shouldn't be needed. I originally had some issue when compiling without this trick. It seems fine now, I don't know what was happening previously

vepadulano commented on tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:
```diff
@@ -1647,6 +1647,24 @@ public:
       return [this](unsigned int, const RSampleInfo &) mutable { fBranchAddressesNeedReset = true; };
    }
+
+   SnapshotHelper MakeNew(void *newName)
```
Split this change in 2 commits in the most recent update. Also, implemented the `MakeNew` method for `SnapshotHelperMT`

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RResultMap.hxx:
```diff
@@ -52,6 +52,35 @@ MakeResultMap(std::shared_ptr<T> nominalResult, std::vector<std::shared_ptr<T>>
+ *
+ * \tparam T The type of the results held by the map.
+ * \param inmap The map.
+ * \return ROOT::RDF::Experimental::RResultMap<T> A new map with the clone variations.
```
Too much star wars probably

vepadulano commented on tree/dataframe/inc/ROOT/RResultPtr.hxx:
```diff
@@ -41,7 +41,35 @@ class RInterface;
+   // For now this is an extra copy, best would be calling default constructor.
+   // Main reason (as far as I am aware) is copying metadata of certain
+   // result types, e.g. a for a TH1D we have to create a new histogram with
+   // the same binning and axis limits.
```
Fair enough, reworded!

vepadulano commented on tree/dataframe/src/RResultPtr.cxx:
```diff
@@ -0,0 +1,30 @@
+{
+   return SnapshotPtr_t{
+      inptr.fObjPtr, inptr.fLoopManager,
+      inptr.fActionPtr->CloneAction(reinterpret_cast<void *>(const_cast<std::string *>(&newNameForSnapshot)))};
```
Not really unfortunately, cast to `void *` discards qualifier and the compiler complains, I need the `const_cast` (or a cast to `const void *`, but that would clash with the signature of `MakeNew(void *)`)

eguiraud replied:
ah right, then only the other side can be simplified: you can directly cast `void *` to `const std::string *`

vepadulano commented on bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py:
```diff
@@ -241,3 +241,34 @@ def _(mergeable: ROOT.Detail.RDF.RMergeableVariationsBase, node, backend):
+
+
+@clone_action.register(AsNumpyResult)
+def _(asnumpyres):
```
In the latest update, the logic is implemented fully via `singledispatch`

vepadulano commented:
Thank you so much @eguiraud for the comments! I have addressed all of them and updated the commits accordingly, also splitting them in smaller commits to make intentions clearer.

New tests are still missing, but I would appreciate if you could take another look. In particular, these changes hopefully streamline the code a bit:
* `rdf_uuid`: this was previously not well named, since we need something that contains both an identifier for the current RDataFrame instance and also one for the current (distributed) execution of that RDataFrame (e.g. to separate to consecutive triggers of the computation graph). I have added the type `ExecutionIdentifier` that holds both. This class is now used everywhere (in Ranges.py, HeadNode.py) and should improve readability.
* I merged `CloneResultPtr` and `CloneResultMap` into `CloneResultAndAction`. This helps making the logic in DistRDF more simple (i.e. is what allows the usage of `singledispatch` to distinguish different ways of cloning actions of the distributed computation graph).

eguiraud commented:
Thank you @vepadulano !

I'm happy to do another pass, but I'd rather do so after the tests are added: it takes me around half a day of work to go through a PR this size, I'd like to reduce the amount of passes... :sweat_smile: 

(the warnings are real and the test failure is because of an expected error message that needs to be updated in the test -- or the new message needs to match the old one)

vepadulano commented:
This now depends on #12363 for the failures on ubuntu18

vepadulano commented:
Added tests for the C++ mechanisms, and attempt at fixing the ubuntu18 failure with a much more simplified approach w.r.t. https://github.com/root-project/root/pull/12981, which should not be needed anymore (let's see what the CI thinks).

vepadulano commented:
@phsft-bot build

vepadulano commented:
Added tests on the distributed RDF side too.

@eguiraud this is now ready for review again

eguiraud commented on bindings/experimental/distrdf/python/DistRDF/PythonMergeables.py:
```diff
@@ -48,3 +49,7 @@ def GetValue(self, backend: BaseBackend):
+
+
+# A type alias to signify any type of result that can be returned from the RDataFrame API
+RDataFrameFutureResult = Union[ROOT.RDF.RResultPtr, ROOT.RDF.Experimental.RResultMap, SnapshotResult, AsNumpyResult]
```
not for this PR, but maybe `AsNumpyResult` should live in this module and be imported by `_pythonization._rdataframe` instead?

vepadulano replied:
Unclear, `AsNumpyResult` is actually tied to the `AsNumpy` feature, before being used in `DistRDF`. Moving it here would mean making that part of PyROOT depend on a class defined in a `DistRDF` submodule. One alternative is splitting the class in two. The actual result of `AsNumpy` would still live only in PyROOT, whereas in `DistRDF` I could create a second class and call it `AsNumpyFuture` or similar, that wraps the `AsNumpyResult` from PyROOT and then adds the `Merge` functionality on top

eguiraud commented on bindings/experimental/distrdf/python/DistRDF/_graph_cache.py:
```diff
@@ -0,0 +1,30 @@
+from __future__ import annotations
```
I guess we don't need this, if we are using dataclasses we assume at least Python 3.7

vepadulano replied:
Delayed evaluation of annotations is a Python 3.7 feature indeed, but needs to be enforced via the `__future__` import. Unfortunately, the Python standard hasn't yet decided on whether this will actually be the default behaviour (see https://docs.python.org/3/library/__future__.html#id1). In DistRDF we use it extensively both to avoid circular imports and to avoid importing modules that are used only for annotations, so I just write the `__future__` import in every module

eguiraud commented on bindings/experimental/distrdf/python/DistRDF/_graph_cache.py:
```diff
@@ -0,0 +1,30 @@
+    from DistRDF.PythonMergeables import RDataFrameFutureResult
+
+
+@dataclass(eq=True, frozen=True)
```
isn't `eq=True` the default?

eguiraud commented on bindings/experimental/distrdf/python/DistRDF/Ranges.py:
```diff
@@ -157,6 +128,51 @@ class TaskTreeEntries:
+def get_balanced_ranges(nentries, npartitions, exec_id: ExecutionIdentifier):
+    """
+    Builds range pairs from the given values of the number of entries in
+    the dataset and number of partitions required. Each range contains the
+    same amount of entries, except for those cases where the number of
+    entries is not a multiple of the partitions.
```
maybe more simply you can just say "The `nentries` are divided uniformly among the `npartitions`".

eguiraud commented on tree/dataframe/test/dataframe_cloning.cxx:
```diff
@@ -0,0 +1,405 @@
+   for (const auto &name : fileNames) {
+      gSystem->Unlink(name.c_str());
+   }
+}
```
NICE

eguiraud reviewed (approved):
LGTM! Test coverage looks great and the new commit splitting makes a lot of sense.

Only minor comments, feel free to ignore.

vepadulano commented on tree/dataframe/src/RLoopManager.cxx:
```diff
@@ -369,20 +369,31 @@ RLoopManager::RLoopManager(std::unique_ptr<RDataSource> ds, const ColumnNames_t
+   ChangeSpec(std::move(spec));
+}
+
+void RLoopManager::ChangeSpec(ROOT::RDF::Experimental::RDatasetSpec &&spec)
```
Indeed, the new spec is expected to provide exactly the same dataset schema. I agree with you that there is a very dangerous interaction potentially. We can't really do otherwise for now. I will add some very explicit docs in the next commit

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RVariedAction.hxx:
```diff
@@ -194,6 +194,22 @@ public:
       throw std::logic_error("Cannot produce a varied action from a varied action.");
    }
 
+   std::unique_ptr<RActionBase> CloneAction(void *typeErasedResults) final
+   {
+      const auto &vectorOfTypeErasedResults = *reinterpret_cast<const std::vector<void *> *>(typeErasedResults);
```
I don't think that's possible since `RActionBase` declares
```
virtual std::unique_ptr<RActionBase> CloneAction(void *newResult) = 0;
```

vepadulano replied:
I mean, I could always make a `RVariedAction::CloneActionImpl(const std::vector<void *> &results)` and then `RVariedAction::CloneAction(void *results)` would just reinterpret cast the pointer to a vector and call the `Impl` function, but that's not really necessary imho
